Feature: experimental meson buildsystem (#1528)
blackmagic-debug/blackmagic: src/platforms/hosted/README.md

# Black Magic Debug App (BMDA)

Compile with:

```sh
meson setup build
meson compile -C build

> ```

Reviewer: It'd be more readable to remove this and integrate it with Description as shown below.

## Description

> BMDA runs on your computer and compiles as `blackmagic`.

Reviewer: ```suggestion
BMDA can be compiled as follows:

```sh
meson setup build
meson compile -C build
```

```

> When started, it either presents a list of available probes or starts a GDB server on
> localhost port 2000 if possible (if there's already something bound, it tries the next
> 4 port numbers up to 2004), and if there is either only one probe is attached to the
> computer or enough information is given on the command line to select one of several probes.

Reviewer: ```suggestion
When the `blackmagic` executable is launched, it will present a list of available probes.
Alternatively, if there is only one probe attached to the computer, or enough information is given on the
command line to select one of several probes, it will start a GDB server on localhost port 2000
(if the port is already bound, BMDA will try the next 4 port numbers up to 2004).
```

You can connect to the server as you would connect to the BMP over USB serial, replacing
the device path with `:2000` as in `target extended-remote :2000`. BMDA's GDB functionality
is the same, but the monitor options are a little different and can be viewed with `monitor help`.

The following lists some of the supported additional commmand line options and what they do.

> It is not exhaustive, and does not show  how they can all be combined.

Reviewer: ```suggestion
It is not exhaustive, and does not show how they can all be combined.
```

### Display the complete program usage help on the console

```sh
blackmagic -h
```

### Run "Test Mode" and display any connected targets

For SWD:

```sh
blackmagic -t
```

For JTAG:

```sh
blackmagic -tj
```

### Directly flash a binary file at lowest flash address

```sh
blackmagic <file>
```

or with the -S argument at some other address

```sh
blackmagic -S 0x08002000 <file>
```

NB: BMDA will treat whatever file you provide here as the raw contents to write to Flash.

> It does not understand ELF, Intel Hex, or SREC formats.

Reviewer: Move these lines to the beginning of the section and **bolden** them.

### Read flash to binary file

```sh
blackmagic -r <file>
```

### Verify flash against binary file

```sh
blackmagic -V <file>
```

### Show available monitor commands

```sh
blackmagic -M help
```

### Show available monitor commands for the second target found

```sh
blackmagic -n 2 -M help
```

### Monitor commands with multiple arguments - e.g. for STM32 option bytes

```sh
blackmagic -M "option help"
```

## Dependencies for non-`HOSTED_BMP_ONLY=1` builds

BMDA uses the following external libraries to function when built in full:

* libusb1
* libftdi1 (except on Windows)
* hidapi (hidapi-hidraw for Linux)

> ## Compiling on windows

Reviewer: ```suggestion
## Compiling on Windows
```

# PC-Hosted BMP
Compile with:

> ``` sh

Reviewer 2: These all need going through and correcting slightly - there should not be a space between the last backtick and "sh".

> cd blackmagic

Reviewer 2: Similarly for this

meson setup build
cd build
meson compile
```
This needs minimal external support.

> TODO: update this, meson currectly does not support "HOSTED_BMP_ONLY"

Reviewer 2: This is intentional as we're interested in entirely eliminating this build mode by fully statically linking all the dependencies into the BMDA binary so there's never a reason to do a BMP-only build.

Author: Ah, good to know, removed.

you may get compilation errors. You will need to install the libusb
and libftdi libraries, and have the correct mingw compiler.
You can use these commands to install dependencies, and build PC-hosted BMP

> from a mingw64 shell, from within the `blackmagic` directory:

Reviewer 2: This should be `UCRT64 shell`, and should use `pacboy` - This entire passage is outdated.

You could reasonable link the [build instructions for Windows](https://black-magic.org/knowledge/compiling-windows.html#building-bmda-on-a-windows-machine) from the website as we rewrote all this there.

Author: Hm, I'm not comfortable with the windows workflow, do you mind writing it for me? I can apply it if you just give me an excerpt

Reviewer 2: Sure thing, we'll review the passage fully and write a suggestion a bit later today then.
